### PR TITLE
[DISCUSSION NEEDED] AV-sync fundamental issues

### PR DESCRIPTION
**Pull request by bjuncek**
As pointed out in #1221 `_read_from_stream` function assumes that all the offsets are global, which is fundamentally wrong as `pts` is stream-specific representation. 

Here, I propose to make precomputed `pts` a global offset in the form of `k/fps` and type `fraction.Fraction` where `k` is k-th frame. Then, in every function which operates on stream level, global `pts` would be converted to a stream specific measure by dividing it by stream specific time-base, i.e. `int(round(float(global_offset / stream_time_base)))`.

Obviously, tests are modified to reflect this change.

--- 

This is an early version and a suggestion on how to tackle formerly mentioned issue. If anyone has better Idea, please let me know. 

cc @fmassa @iyah4888

### CODE REVIEW

File: torchvision/io/video.py [hunk 1]
```diff
@@ -232,6 +248,8 @@ def read_video_timestamps(filename):
     -------
     pts : List[int]
         presentation timestamps for each one of the frames in the video.
+        Note that these are returned as the TRUE timestamps, i.e.
```
Comment (fmassa) [inline torchvision/io/video.py]: can you mention `the true timestamps in seconds`?

File: torchvision/io/video.py [hunk 2]
```diff
@@ -129,10 +139,16 @@ def _read_from_stream(container, start_offset, end_offset, stream, stream_name):
         # we will have all the necessary data. This is most useful for audio
         first_frame_pts = max(i for i in frames if i < start_offset)
         result.insert(0, frames[first_frame_pts])
+    # print(f'Seek off {seek_offset}, end off {end_offset}, result len {len(result)}')
```
Comment (fmassa) [inline torchvision/io/video.py]: can you remove this comment?

File: test/test_io.py [hunk 3]
```diff
@@ -76,39 +77,40 @@ def test_read_timestamps(self):
...
+            pts, fps = io.read_video_timestamps(f_name)
             for start in range(5):
                 for l in range(1, 4):
                     lv, _, _ = io.read_video(f_name, pts[start], pts[start + l - 1])
                     s_data = data[start:(start + l)]
                     self.assertEqual(len(lv), l)
                     self.assertTrue(s_data.equal(lv))
 
-            lv, _, _ = io.read_video(f_name, pts[4] + 1, pts[7])
+            lv, _, _ = io.read_video(f_name, pts[4] + 1 / (fps + 1), pts[7])
```
Comment (fmassa) [inline test/test_io.py]: Can you explain the need of the `+1` in `(fps + 1)`?

File: test/test_io.py [hunk 4]
```diff
@@ -138,13 +140,32 @@ def test_read_timestamps_from_packet(self):
...
+                pts, fps = io.read_video_timestamps(f_name)
+                self.assertEqual(pts, sorted(pts))
+                self.assertEqual(fps, 30)
+            except URLError:
+                msg = "could not download test file '{}'".format(url)
+                warnings.warn(msg, RuntimeWarning)
+                raise unittest.SkipTest(msg)
+
+            lv, la, info = io.read_video(f_name, pts[3], pts[7])
+            # FIXME: add Another video - this one doesn't have audio
```
Comment (fmassa) [inline test/test_io.py]: Another option for getting the video would be to use `av.datasets` https://github.com/mikeboers/PyAV/blob/cd458ffe89988b0feca44da6c56ef29f17555962/tests/common.py#L11

File: torchvision/io/video.py [hunk 5]
```diff
@@ -250,4 +268,4 @@ def read_video_timestamps(filename):
                                              container.streams.video[0], {'video': 0})
         video_fps = float(container.streams.video[0].average_rate)
     container.close()
-    return [x.pts for x in video_frames], video_fps
+    return [x.pts * x.time_base for x in video_frames], video_fps
```
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/io/video.py]: Yeah, that makes a lot of sense.
Is there a preferred warning system in torchvision?

File: torchvision/io/video.py (hunk 1 above)
Comment (bjuncek) [reply torchvision/io/video.py]: yup.

File: test/test_io.py (hunk 4 above)
Comment (bjuncek) [reply test/test_io.py]: Sounds reasonable  - I'll take a look.

File: torchvision/io/video.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fmassa) [reply torchvision/io/video.py]: let's just use `warnings.warn` for now

File: test/test_io.py (hunk 3 above)
Comment (bjuncek) [reply test/test_io.py]: My understanding was that the original test was here to make sure that if we try to decode from a pts that doesn't exist (`pts[4]+1`) that we return the closest possible frame to that pts - then we check and get 4 frames (from `pts[4]` to `pts[7]`).

In the same way, `1/(fps+1)` is a non-existing pts that is the closest to the existing (`pts[4]`).

File: torchvision/io/video.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bjuncek) [reply torchvision/io/video.py]: S.G.
what do we want to keep as a default behaviour?

**Comment by fmassa** [comment]
Subsumed by https://github.com/pytorch/vision/pull/1331

**Pull request closed without merging**